### ISSUE TITLE
review how we duck-type non-numpy inputs

### ISSUE THREAD

**Issue by tacaswell** [matplotlib/matplotlib#16402]
We currently do a fair amount of ad-hoc normalization / edge case handling for when we get pandas (or xarray or TF or ...) objects that are _close_ to numpy but not quite (for example https://github.com/matplotlib/matplotlib/pull/16347 which uses a warning context manager which is extra not-thread-safe in other places we check of `obj.values`).

Can we centralize / streamline this logic?

**Comment by tacaswell** [comment]
also xref: https://github.com/matplotlib/matplotlib/issues/16400

**Comment by jklymak** [comment]
We don't really ever get complaints from TF or xarray, so the practical question is how much do we special-case pandas, and who will set up the tests, and lead the charge on integration? This is, in my opinion, hampered by not having anyone on the lead-dev team who regularly uses pandas.  It would be nice to try and entrain someone who did, who could advocate for changes and/or interface with the pandas project (who I have found are great to work with, but you know, are pushing their own things forward).

**Comment by timhoffm** [comment]
`.values` is not recommended anymore in pandas: https://pandas.pydata.org/docs/reference/api/pandas.Series.values.html / https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.values.html. Probably `to_numpy()` is the way to go.

**Comment by jklymak** [comment]
Yeah, unfortunately, thats not universal.  `xarray` still uses `.values`.

**Comment by oscargus** [comment]
Which libraries are we talking about? pandas, xarray, polars, tensorflow(?), ...

I've added a test for xarray in #22560, which also centralizes this, but probably makes sense to add tests for all libraries that is of interest (there was one for pandas). (polars can not yet be installed with conda.)

**Comment by timhoffm** [comment]
I close this. #22879 is a more recent discussion, and I think there are others as well.
